File: evd_ros_backend/evd_ros_core/src/evd_script/program/primitives/move_trajectory.py
'''
Moves a robot according to a preplanned trajectory. This wraps the trajectory
data structure with additional movement parameterization needed to actually
execute on the robot.

TODO implement real-time move trajectory behavior
TODO implement thing token movement behavior
'''

from ..primitive import Primitive
from ...data.trajectory import Trajectory
from ...node_parser import NodeParser

class MoveTrajectory(Primitive):

    '''
    Data structure methods
    '''

    @classmethod
    def type_string(cls, trailing_delim=True):
        return 'move-trajectory' + ('.' if trailing_delim else '')

    @classmethod
    def full_type_string(cls):
        return Primitive.full_type_string() + cls.type_string()

    def __init__(self, startLocUuid=None, endLocUuid=None, trajectory=None, trajectory_uuid=None,
                 type='', name='', uuid=None, parent=None, append_type=True):

        self._context_patch = None
        self._start_location_uuid = None
        self._end_location_uuid = None
        self._trajectory_uuid = None

        super(MoveTrajectory,self).__init__(
            type=MoveTrajectory.type_string() + type if append_type else type,
            name=name,
            uuid=uuid,
            parent=parent,
            append_type=append_type)

        self.start_location_uuid = startLocUuid
        self.end_location_uuid = endLocUuid

        if trajectory != None and trajectory_uuid != None:
            raise Exception('Cannot supply both a default trajectory and a trajectory id already in context')
        elif trajectory != None:
            self.trajectory = trajectory
        elif trajectory_uuid != None:
            self.trajectory_uuid = trajectory_uuid
        else:
            self.trajectory = Trajectory(self.start_location_uuid,self.end_location_uuid)

    def to_dct(self):
        tmp = self.context # This guarantees that the patch is applied if possible

        msg = super(MoveTrajectory,self).to_dct()
        msg.update({
            'start_location_uuid': self.start_location_uuid,
            'end_location_uuid': self.end_location_uuid,
        })

        if self._context_patch == None:
            msg['trajectory_uuid'] = self.trajectory_uuid
        else:
            msg['trajectory'] = self.trajectory.to_dct()

        return msg

    @classmethod
    def from_dct(cls, dct):
        return cls(
            name=dct['name'],
            type=dct['type'],
            append_type=False,
            uuid=dct['uuid'],
            startLocUuid=dct['start_location_uuid'],
            endLocUuid=dct['end_location_uuid'],
            trajectory=NodeParser(dct['trajectory'], enforce_type=Trajectory.type_string(trailing_delim=False)) if 'trajectory' in dct.keys() else None,
            trajectory_uuid=dct['trajectory_uuid'] if 'trajectory_uuid' in dct.keys() else None)

    def on_delete(self):
        self.context.delete_trajectory(self.trajectory_uuid)

        super(MoveTrajectory,self).on_delete()

    '''
    Data accessor/modifier methods
    '''

    @property
    def context(self):
        # Need to patch context when none exists
        realContext = super(MoveTrajectory,self).context
        if realContext == None:
            if self._context_patch == None:
                self._context_patch = ContextPatch(self)
            return self._context_patch
        else:
            if self._context_patch != None:
                self._context_patch.update_context(realContext)
                self._context_patch = None
            return realContext

    @property
    def start_location_uuid(self):
        return self._start_location_uuid

    @start_location_uuid.setter
    def start_location_uuid(self, value):
        if self._start_location_uuid != value:
            self._start_location_uuid = value

            t = self.context.get_trajectory(self.trajectory_uuid)
            if t != None:
                t.start_location_uuid = self._start_location_uuid

            self.updated_attribute('start_location_uuid','set')

    @property
    def end_location_uuid(self):
        return self._end_location_uuid

    @end_location_uuid.setter
    def end_location_uuid(self, value):
        if self._end_location_uuid != value:
            self._end_location_uuid = value

            t = self.context.get_trajectory(self.trajectory_uuid)
            if t != None:
                t.end_location_uuid = self._end_location_uuid

            self.updated_attribute('end_location_uuid','set')

    @property
    def trajectory(self):
        return self.context.get_trajectory(self.trajectory_uuid)

    @trajectory.setter
    def trajectory(self, value):
        if self._trajectory_uuid != value.uuid:
            self.context.delete_trajectory(self.trajectory_uuid)
            self.context.add_trajectory(value)
            self.trajectory_uuid = value.uuid

            self.updated_attribute('trajectory','set')

    @property
    def trajectory_uuid(self):
        return self._trajectory_uuid

    @trajectory_uuid.setter
    def trajectory_uuid(self, value):
        if self._trajectory_uuid != value:
            if value == None:
                raise Exception('Id must be defined')

            traj = self.context.get_trajectory(value)
            if traj == None and self._context_patch == None:
                raise Exception('Id must have a trajectory already in context')

            elif traj == None and self._context_patch != None:
                self._context_patch.add_pending_trajectory(
                    uuid=value,
                    startLoc=self.start_location_uuid,
                    endLoc=self.end_location_uuid)

            elif traj != None:
                if traj.start_location_uuid != self.start_location_uuid:
                    raise Exception('Trajectory must have matching start location to primitive if set by id')

                if traj.end_location_uuid != self.end_location_uuid:
                    raise Exception('Trajectory must have matching end location to primitive if set by id')

            self._trajectory_uuid = value
            self.updated_attribute('trajectory_uuid','set')

    def set(self, dct):
        if 'start_location_uuid' in dct.keys():
            self.start_location_uuid = dct['start_location_uuid']

        if 'end_location_uuid' in dct.keys():
            self.end_location_uuid = dct['end_location_uuid']

        if 'trajectory' in dct.keys():
            self.trajectory = NodeParser(dct['trajectory'], enforce_type=Trajectory.type_string(trailing_delim=False))

        if 'trajectory_uuid' in dct.keys():
            self.trajectory_uuid = dct['trajectory_uuid']

        super(MoveTrajectory,self).set(dct)

    '''
    Update Methods
    '''

    def late_construct_update(self):
        tmp = self.context # This guarantees that the patch is applied if possible

        super(MoveTrajectory,self).late_construct_update()

    def deep_update(self):
        tmp = self.context # This guarantees that the patch is applied if possible

        super(MoveTrajectory,self).deep_update()

        self.updated_attribute('start_location_uuid','update')
        self.updated_attribute('end_location_uuid','update')
        self.updated_attribute('trajectory_uuid','update')
        self.updated_attribute('trajectory','update')

    def shallow_update(self):
        super(MoveTrajectory,self).shallow_update()

        self.updated_attribute('start_location_uuid','update')
        self.updated_attribute('end_location_uuid','update')
        self.updated_attribute('trajectory_uuid','update')
        self.updated_attribute('trajectory','update')

    '''
    Execution methods
    '''

    def symbolic_execution(self, hooks):
        hooks.active_primitive = self

        loc = self.context.get_location(self.end_location_uuid)
        hooks.tokens['robot']['state']['position'] = loc.position.to_simple_dct()
        hooks.tokens['robot']['state']['orientation'] = loc.orientation.to_simple_dct()
        hooks.tokens['robot']['state']['joints'] = loc.joints

        return self.parent

    def realtime_execution(self, hooks):

        # TODO actually execute move trajectory behavior

        return self.symbolic_execution(hooks)


class ContextPatch(object):

    def __init__(self, parent):
        self._trajectories = {}
        self._parent = parent
        self._pending = {}

    @property
    def parent(self):
        return self._parent

    @property
    def trajectories(self):
        traj = {uuid: self._trajectories[uuid] for uuid in self._trajectories.keys()} # copy
        traj.update({uuid: 'pending' for uuid in self._pending.keys()})
        return traj

    @trajectories.setter
    def trajectories(self, value):
        for t in self._trajectories:
            t.remove_from_cache()
        self._trajectories = {}

        for t in value:
            self._trajectories[t.uuid] = t
            t.parent = self

    def get_trajectory(self, uuid):
        if uuid in self._trajectories.keys():
            return self._trajectories[uuid]
        elif uuid in self._pending.keys():
            return 'pending'
        else:
            return None

    def add_trajectory(self, trajectory):
        trajectory.parent = self
        self._trajectories[trajectory.uuid] = trajectory

    def delete_trajectory(self, uuid):
        if uuid in self._trajectories.keys():
            self._trajectories.pop(uuid).remove_from_cache()
        elif uuid in self._pending.keys():
            self._pending.pop(uuid)

    def update_context(self, context):

        for traj in self._trajectories.values():
            context.add_trajectory(traj)

        for uuid in self._pending.keys():
            traj = context.get_trajectory(uuid)
            if traj == None:
                raise Exception('Pending trajectory was not found when context was assigned')

            if traj.start_location_uuid != self._pending[uuid]['startLoc']:
                raise Exception('Pending trajectory start location does not match one in context')

            if traj.end_location_uuid != self._pending[uuid]['endLoc']:
                raise Exception('Pending trajectory end location does not match one in context')

        self._trajectories = {}
        self._pending = {}

    def child_changed_event(self, attribute_trace):
        self.parent.child_changed_event(attribute_trace)

    def add_pending_trajectory(self, uuid, startLoc, endLoc):
        self._pending[uuid] = {'startLoc': startLoc, 'endLoc': endLoc}

File: evd_authoring_app/src/model/evd_script/program/primitives/move_trajectory.js
import { Primitive } from '../primitive';
import { Trajectory } from '../../data/trajectory';
import { Context } from '../../context';


export class MoveTrajectory extends Primitive {

    /*
    * Data structure methods
    */

    static typeString() {
        return 'move-trajectory.';
    }

    static fullTypeString() {
        return Primitive.fullTypeString() + MoveTrajectory.typeString();
    }

    constructor(startLocUuid=null, endLocUuid=null, trajectory=null, trajectoryUuid=null, type='', name='', uuid=null, parent=null, appendType=true) {
        this._contextPatch = null;
        this._startLocationUuid = null;
        this._endLocationUuid = null;
        this._trajectoryUuid = null;

        super(
            type= (appendType) ? 'move-trajectory.'+type : type,
            name= name,
            uuid= uuid,
            parent= parent,
            appendType= appendType
        );

        this.startLocationUuid = startLocUuid;
        this.endLocationUuid = endLocUuid;

        if (trajectory !== null && trajectoryUuid !== null) {
            throw new Error('Cannot supply both a default trajectory and a trajectory id already in context');
        } else if (trajectory !== null) {
            this.trajectory = trajectory;
        } else if (trajectoryUuid !== null) {
            this.trajectoryUuid = trajectoryUuid;
        } else {
            this.trajectory = new Trajectory(this.startLocationUuid,this.endLocationUuid);
        }
    }

    toDict() {
        const tmp = this.context; // force a context update before serialization

        let msg = {
            ...super.toDict(),
            start_location_uuid: this.startLocationUuid,
            end_location_uuid: this.endLocationUuid
        };

        if (this._contextPatch === null) {
            msg.trajectory_uuid = this.trajectoryUuid;
        } else {
            msg.trajectory = this.trajectory.toDict();
        }

        return msg;
    }

    static fromDict(dct) {
        return new MoveTrajectory(
            name= dct.name,
            type= dct.type,
            appendType= false,
            uuid= dct.uuid,
            startLocUuid= dct.start_location_uuid,
            endLocUuid = dct.end_location_uuid,
            trajectory= ('trajectory' in dct) ? Trajectory.fromDict(dct.trajectory) : null,
            trajectoryUuid= ('trajectory_uuid' in dct) ? dct.trajectory_uuid : null 
        );
    }

    /*
    * Accessor/modifier methods
    */

    context() {
        // Need to patch context when none exists
        const realContext = super.context;
        if (realContext === null) {
            if (this._contextPatch === null) {
                this._contextPatch = new ContextPatch();
            }
            return this._contextPatch;
        } else {
            if (this._contextPatch !== null) {
                this._contextPatch.updateContext(realContext);
                this._contextPatch = null;
            }
            return realContext;
        }
    }

    get startLocationUuid() {
        return this._startLocationUuid;
    }

    set startLocationUuid(value) {
        if (this._startLocationUuid !== value) {
            this._startLocationUuid = value;

            const t = this.context.getTrajectory(this.trajectoryUuid);
            if (t !== null) {
                t.startLocationUuid = this._startLocationUuid;
            }

            this.updatedAttribute('start_location_uuid','set');
        }
    }

    get endLocationUuid() {
        return this._endLocationUuid;
    }

    set endLocationUuid(value) {
        if (this._endLocationUuid !== value) {
            this._endLocationUuid = value;

            const t = this.context.getTrajectory(this.trajectoryUuid);
            if (t !== null) {
                t.endLocationUuid = this._endLocationUuid;
            }

            this.updatedAttribute('end_location_uuid','set');
        }
    }

    get trajectory() {
        return this.context.getTrajectory(this.trajectoryUuid);
    }

    set trajectory(value) {
        if (this._trajectoryUuid !== value.uuid) {
            this.context.deleteTrajectory(this.trajectoryUuid);
            this.context.addTrajectory(value);
            this.trajectoryUuid = value.uuid;
        }
    }

    get trajectoryUuid() {
        return this._trajectoryUuid;
    }

    set trajectoryUuid(value) {
        if (this.trajectoryUuid !== value) {
            if (value === null) {
                throw new Error('Id must be defined');
            }

            const traj = this.context.getTrajectory(value);
            if (traj === null && this._contextPatch === null) {
                throw new Error('Id must have a trajectory already in context');
            } else if (traj === null && this._contextPatch !== null) {
                this._contextPatch.addPendingTrajectory(
                    uuid=value,
                    startLoc=this.startLocationUuid,
                    endLoc=this.endLocationUuid
                );
            } else if (traj !== null) {
                if (traj.startLocationUuid !== this.startLocationUuid) {
                    throw new Error('Trajectory must have matching start location to primtive if set by id');
                }

                if (traj.endLocationUuid !== this.endLocationUuid) {
                    throw new Error('Trajectory must have matching end location to primitive if set by id');
                }
            }

            this._trajectoryUuid = value;
            this.updatedAttribute('trajectory_uuid','set');
        }
    }

    set(dct) {

        if ('start_location_uuid' in dct) {
            this.startLocationUuid = dct.start_location_uuid;
        }

        if ('end_location_uuid' in dct) {
            this.endLocationUuid = dct.end_location_uuid;
        }

        if ('trajectory' in dct) {
            this.trajectory = Trajectory.fromDict(dct.trajectory);
        }

        if ('trajectory_uuid' in dct) {
            this.trajectoryUuid = dct.trajectory_uuid;
        }

        super.set(dct);
    }

    /*
    * Update methods
    */

    lateConstructUpdate() {
        const tmp = this.context; // force update context;

        super.lateConstructUpdate();
    }

    deepUpdate() {
        const tmp = this.context; // force update context

        super.deepUpdate();

        this.updatedAttribute('start_location_uuid','update');
        this.updatedAttribute('end_location_uuid','update');
        this.updatedAttribute('trajectory_uuid','update');
        this.updatedAttribute('trajectory','update');
    }

    shallowUpdate() {
        super.shallowUpdate();

        this.updatedAttribute('start_location_uuid','update');
        this.updatedAttribute('end_location_uuid','update');
        this.updatedAttribute('trajectory_uuid','update');
        this.updatedAttribute('trajectory','update');
    }
}


class ContextPatch {

    constructor(parent) {
        this._trajctories = {};
        this._parent = parent;
        this._pending = {};
    }

    get parent() {
        return this._parent;
    }

    get trajectories() {
        const traj = {
            ...this._trajctories,
            ...this._pending
        };
        return traj;
    }

    set trajectories(value) {
        for (const [uuid, t] of Object.entries(this._trajctories)) {
            t.removeFromCache();
        }
        this._trajctories = {};

        for (const [uuid, t] of Object.entries(value)) {
            this._trajctories[uuid] = t;
            t.parent = this;
        }
    }

    getTrajectory(uuid) {
        if (uuid in this._trajctories) {
            return this._trajctories[uuid];
        } else if (uuid in this._pending) {
            return 'pending';
        } else {
            return null;
        }
    }

    addTrajectory(trajectory) {
        trajectory.parent = this;
        this._trajctories[trajectory.uuid] = trajectory;
    }

    deleteTrajectory(uuid) {
        if (uuid in this._trajctories) {
            this._trajctories[uuid].removeFromCache();
            delete this._trajctories[uuid];
        } else if (uuid in this._pending) {
            delete this._pending[uuid];
        }
    }

    updateContext(context) {

        for (const [uuid, t] of Object.entries(this._trajctories)) {
            context.addTrajectory(t);
        }

        for (const [uuid, obj] of Object.entries(this._pending)) {
            const traj = context.getTrajectory(uuid);
            if (traj === null) {
                throw new Error('Pending trajectory was not found when context was assigned');
            }

            if (traj.startLocationUuid !== obj.startLoc) {
                throw new Error('Pending trajectory start location does not match one in context');
            }

            if (traj.endLocationUuid !== obj.endLoc) {
                throw new Error('Pending trajectory end location does not match one in context');
            }

            this._trajctories = {};
            this._pending = {};
        }
    }

    childChangedEvent(attributeTrace) {
        this.parent.childChangedEvent(attributeTrace);
    }

    addPendingTrajectory(uuid, startLoc, endLoc) {
        this._pending[uuid] = {startLoc, endLoc};
    }

}